File: chap06_waves/bistable_waves_pp.m
% bistable waves  phase plane
%  -------------------------------------------------------------------
%
%  % find the traveling wave solution of the bistable equation using
%  shooting 
%
%   For Chapter 6, Figure 1.8, 1.9 of
%   Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
% 
%   Written by James Keener and James Sneyd
% 
%  ------------------------------------------------------------------- 

 clear all; close all; clc;
 set(0,                           ...
   'defaultaxesfontsize', 20,   ...
   'defaultaxeslinewidth', 2.0, ...
   'defaultlinelinewidth', 2.0, ...
   'defaultpatchlinewidth', 0.7);  

 
global alf c
 
% 
cmykflg=1;
%if cmykflg=1, this code will make cmyk figure files
%cmykflg = 1;

alf = 0.1;
clist = [0;0.56;0.57;1;0.5665984]
for j = 1:5
c = clist(j);

lam = (-c+sqrt(c^2+4*alf))/2;
u0 = 0.001;
w0 = lam*u0;

s = [u0;w0];
tstep = 0.1;
t_end = 25;

tspan = [ 0:tstep:t_end];
[T,S] = ode23s(@deRHS,tspan,s,odeset('Events',@events)); 


figure(1)
if (j<5)
plot(S(:,1),S(:,2),'linewidth',2)
 hold on
 axis([0 1 0 0.2])
else
    plot(S(:,1),S(:,2),'--','linewidth',2)
 
 xlabel('U','fontsize',20)
 ylabel('W','fontsize',20)
end
text(0.08,0.18,'c=1','fontsize',20)
text(0.15,0.02,'c=0','fontsize',20)
text(0.67,0.02,'c=0.56','fontsize',20)
text(0.8,0.14,'c=0.57','fontsize',20)
 box off
end
hold off

 

[tmx,tj] = max(S(:,2));
tshift = T(tj);
figure(2)
plot(T-tshift,S(:,2),T-tshift,S(:,1),'linewidth',2)
text(3.5,0.1,'W(\zeta)','fontsize',20)
text(3.5,0.85,'U(\zeta)','fontsize',20)
 xlabel('\zeta','fontsize',20)
 axis([-10 10 0 1])
 box off
function s_prime=deRHS(t,s)
global alf c
v = s(1);
w = s(2);
 
Fv = w;
Fw = c*w-v*(1-v)*(v-alf);

s_prime = [Fv Fw]';
end

% event function
function [value,isterminal,direction]=events(t,s)
v = s(1);
w = s(2);

value = w*(1-v); 
isterminal = 1;
direction = 0;
end

File: chap06_waves/bistable_waves_pp.py
#   -------------------------------------------------------------------
# 
#   Plot some solutions of the bistable equation for different values of 
#   c. A demonstration of how shooting can work to find the wave speed.
#
#    The wave speed is slightly different from that calculated by the Matlab code, 
#    but closer to the correct value of (1-2*alp)/(sqrt(2)) = 0.56568542.
# 
#    For Chapter 6, Figure 6.3 of
#    Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
# 
#    Written by James Keener and James Sneyd
# 
#   -------------------------------------------------------------------

import numpy as np
from scipy.integrate import solve_ivp
import matplotlib.pyplot as plt


def de_rhs(t, s):
    v = s[0]
    w = s[1]

    Fv = w
    Fw = c * w - v * (1 - v) * (v - alf)
    return [Fv, Fw]


# Set plot defaults
plt.rcParams.update({'font.size': 20, 'axes.linewidth': 2.0, 'lines.linewidth': 2.0, 'patch.linewidth': 0.7})

alf = 0.1
clist = [0, 0.56, 0.57, 1, 0.5657133]
tendlist = [50,25,25,10,25]

for j in range(5):
    c = clist[j]

    lam = (-c + np.sqrt(c ** 2 + 4 * alf)) / 2
    u0 = 0.0001
    w0 = lam * u0

    s = [u0, w0]
    tstep = 0.1
    t_end = tendlist[j]
    tspan = np.arange(0, t_end + tstep, tstep) 
    sol = solve_ivp(de_rhs, [0, t_end], s, t_eval=tspan, method='Radau',rtol=1e-10,atol=1e-10)

    S = sol.y.T
    T = sol.t

    plt.figure(1)
    if j < 4:
        plt.plot(S[:, 0], S[:, 1], linewidth=2)
        plt.axis([0, 1, 0, 0.2])
    else:
        plt.plot(S[:, 0], S[:, 1], '--', linewidth=2)

    if j == 4:
        plt.xlabel('U')
        plt.ylabel('W')

plt.text(0.08, 0.18, 'c=1')
plt.text(0.15, 0.02, 'c=0')
plt.text(0.67, 0.02, 'c=0.56')
plt.text(0.8, 0.14, 'c=0.57')


tmx = np.max(S[:, 1])
tj = np.argmax(S[:, 1])
tshift = T[tj]

plt.figure(2)
plt.plot(T - tshift, S[:, 1], T - tshift, S[:, 0], linewidth=2)
plt.text(3.5, 0.1, 'W(ζ)')
plt.text(3.5, 0.85, 'U(ζ)')
plt.xlabel('ζ')
plt.axis([-10, 10, 0, 1])



